tests/agreement.rs

```diff
@@ -0,0 +1,196 @@
+//! Tests of the Binary Byzantine Agreement protocol.
+//!
+//! There are three properties that are tested:
+//!
+//! - Agreement: If any correct node outputs the bit b, then every correct node outputs b.
+//!
+//! - Termination: If all correct nodes receive input, then every correct node outputs a bit.
+//!
+//! - Validity: If any correct node outputs b, then at least one correct node received b as input.
+//!
+//! TODO: Implement adversaries and send BVAL messages at different times.
+
+extern crate hbbft;
+#[macro_use]
+extern crate log;
+extern crate env_logger;
+
+use std::collections::{BTreeMap, VecDeque};
+
+use hbbft::agreement::{Agreement, AgreementMessage};
+
+#[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+struct NodeId(usize);
+
+/// Message pertaining to a particular Agreement instance.
+type InstanceMessage = (NodeId, AgreementMessage);
+type MessageQueue = VecDeque<InstanceMessage>;
+
+struct TestNode {
+    /// Sender ID.
+    id: NodeId,
+    /// Map from proposer IDs into agreement instances.
+    agreements: BTreeMap<NodeId, Agreement<NodeId>>,
```

Comment:
I'd actually just test a single instance here: All the dispatching based on the proposer ID is implemented here in the test and not in the production code, so all this does is running several agreement tests in parallel.

In the production code, the dispatching is implemented in `common_subset`, so the tests for that module will verify that part of the code.

Reply:
This makes sense. I can revert to testing a single proposer.